### PR TITLE
etcd: use etcd_3_5, etcd_3_3: mark vulnerable to CVE-2023-32082

### PR DESCRIPTION
**Pull request by LeSuisse**
###### Description of changes

etcd 3.3 is end of life for a while and is vulnerable to CVE-2023-32082.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zowoq** [comment]
> The etcd-cluster test is failing due to some cert issues (it was also failing with etcd_3_3).

It was working 10 days ago so this should be addressed. https://hydra.nixos.org/build/219766212

The kubernetes tests need to pass as well.

**Comment by zowoq** [comment]
`etcd-cluster` test is fixed in https://github.com/NixOS/nixpkgs/pull/231307 but it still doesn't work with the 3.5 changes in this PR.

**Comment by LeSuisse** [comment]
PR has been rebased and the `etcd-cluster` fixed. `nixosTests.kubernetes` also succeeds.

File: nixos/tests/etcd-cluster.nix [hunk 1]
```diff
@@ -86,14 +86,16 @@ import ./make-test-python.nix ({ pkgs, ... } : let
...
+      ETCD_CERT_FILE = "${etcd_client_cert}";
+      ETCD_KEY_FILE = "${etcd_client_key}";
+      ETCD_CA_FILE = "${ca_pem}";
+      ETCDCTL_ENDPOINTS = "https://127.0.0.1:2379";
     };
 
     networking.firewall.allowedTCPPorts = [ 2380 ];
   };
+
+  etcdctl_cli = "etcdctl --cacert=${ca_pem} --cert=${etcd_cert} --key=${etcd_key}";
```
Comment (zowoq) [inline nixos/tests/etcd-cluster.nix:98]: Please use env vars for this.

Comment (zowoq) [reply nixos/tests/etcd-cluster.nix:98]:
```diff
diff --git a/nixos/tests/etcd-cluster.nix b/nixos/tests/etcd-cluster.nix
index 8c9b2346922..dadf593456e 100644
--- a/nixos/tests/etcd-cluster.nix
+++ b/nixos/tests/etcd-cluster.nix
@@ -86,6 +86,9 @@ import ./make-test-python.nix ({ pkgs, ... } : let
     };
 
     environment.variables = {
+      ETCDCTL_CACERT = "${ca_pem}";
+      ETCDCTL_CERT = "${etcd_cert}";
+      ETCDCTL_KEY = "${etcd_key}";
       ETCD_CERT_FILE = "${etcd_client_cert}";
       ETCD_KEY_FILE = "${etcd_client_key}";
       ETCD_CA_FILE = "${ca_pem}";
@@ -95,7 +98,6 @@ import ./make-test-python.nix ({ pkgs, ... } : let
     networking.firewall.allowedTCPPorts = [ 2380 ];
   };
 
-  etcdctl_cli = "etcdctl --cacert=${ca_pem} --cert=${etcd_cert} --key=${etcd_key}";
 in {
   name = "etcd-cluster";
 
@@ -136,21 +138,21 @@ in {
         node2.start()
         node1.wait_for_unit("etcd.service")
         node2.wait_for_unit("etcd.service")
-        node2.wait_until_succeeds("${etcdctl_cli} endpoint status")
-        node1.succeed("${etcdctl_cli} put /foo/bar 'Hello world'")
-        node2.succeed("${etcdctl_cli} get /foo/bar | grep 'Hello world'")
+        node2.wait_until_succeeds("etcdctl endpoint status")
+        node1.succeed("etcdctl put /foo/bar 'Hello world'")
+        node2.succeed("etcdctl get /foo/bar | grep 'Hello world'")
 
     with subtest("should add another member"):
-        node1.wait_until_succeeds("${etcdctl_cli} member add node3 --peer-urls=https://node3:2380")
+        node1.wait_until_succeeds("etcdctl member add node3 --peer-urls=https://node3:2380")
         node3.start()
         node3.wait_for_unit("etcd.service")
-        node3.wait_until_succeeds("${etcdctl_cli} member list | grep 'node3'")
-        node3.succeed("${etcdctl_cli} endpoint status")
+        node3.wait_until_succeeds("etcdctl member list | grep 'node3'")
+        node3.succeed("etcdctl endpoint status")
 
     with subtest("should survive member crash"):
         node3.crash()
-        node1.succeed("${etcdctl_cli} endpoint status")
-        node1.succeed("${etcdctl_cli} put /foo/bar 'Hello degraded world'")
-        node1.succeed("${etcdctl_cli} get /foo/bar | grep 'Hello degraded world'")
+        node1.succeed("etcdctl endpoint status")
+        node1.succeed("etcdctl put /foo/bar 'Hello degraded world'")
+        node1.succeed("etcdctl get /foo/bar | grep 'Hello degraded world'")
   '';
 })

```

File: pkgs/servers/etcd/3.3.nix [hunk 2]
```diff
@@ -24,8 +24,6 @@ buildGoPackage rec {
     install -Dm755 bin/* bin/functional/cmd/* -t $out/bin
   '';
 
-  passthru.tests = { inherit (nixosTests) etcd etcd-cluster; };
```
Comment (zowoq) [inline pkgs/servers/etcd/3.3.nix:27]: Need to drop `nixosTests` input from this file as well.

**Comment by zowoq** [comment]
cc @NixOS/kubernetes

**Pull request merged**